linfa-clustering/src/dbscan/algorithm.rs

```diff
@@ -0,0 +1,114 @@
+use crate::dbscan::hyperparameters::DBScanHyperParams;
+use ndarray::{s, Array1, ArrayBase, Axis, Data, Ix1, Ix2};
+use ndarray_stats::DeviationExt;
+
+pub fn dbscan(
+    hyperparameters: &DBScanHyperParams,
+    observations: &ArrayBase<impl Data<Elem = f64>, Ix2>,
+) -> Array1<Option<usize>> {
+    let mut cluster_memberships = Array1::from_elem(observations.dim().1, None);
+    let mut current_cluster_id = 0;
+    for (i, obs) in observations.axis_iter(Axis(1)).enumerate() {
+        if cluster_memberships[i].is_some() {
+            continue;
+        }
+        let neighbors = find_neighbors(&obs, observations, hyperparameters.tolerance());
+        if neighbors.len() < hyperparameters.minimum_points() {
+            continue;
+        }
+        // Now go over the neighbours adding them to the cluster
+        cluster_memberships[i] = Some(current_cluster_id);
+        let mut search_queue = neighbors
+            .iter()
+            .filter(|x| cluster_memberships[[**x]].is_none())
```

Comment:
Given that we always go and access the row corresponding to an index in the neighbors collection, I think it would be interesting to try to return `Vec<(usize, &ArrayBase<impl Data<Elem=64>, Ix1>` from `find_neighbors`.
In that way we should be able to save on a bunch of index accesses, bound checks and so forth.

Reply:
Putting the reference to the `ArrayBase` in the vector means lifetimes are needed. Tried to do it and then I need to add the Sized trait to things etc.

Reply:
```Rust
error[E0038]: the trait `ndarray::data_traits::Data` cannot be made into an object
  --> linfa-clustering/src/dbscan/algorithm.rs:49:1
   |
49 | / fn find_neighbors<'a>(
50 | |     candidate: &'a ArrayBase<impl Data<Elem = f64>, Ix1>,
51 | |     observations: &'a ArrayBase<impl Data<Elem = f64>, Ix2>,
52 | |     eps: f64,
...  |
60 | |     res
61 | | }
   | |_^ the trait `ndarray::data_traits::Data` cannot be made into an object
   |
   = note: the trait cannot require that `Self : Sized`
```

Reply:
Just out of curiosity, could you try using generic notation instead of `impl` notation for that function?

Reply:
Got it! And some improvement, here's the criterion printout:
```
     Running /home/xd009642/code/rust/linfa/target/release/deps/dbscan-abaeb88fc2d9076d
dbscan/10               time:   [580.82 ns 582.51 ns 584.38 ns]                       
                        change: [+1.4560% +2.0619% +2.6529%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) high mild
  3 (3.00%) high severe
dbscan/100              time:   [3.8319 us 3.8423 us 3.8564 us]                        
                        change: [-0.3080% +0.0472% +0.3940%] (p = 0.80 > 0.05)
                        No change in performance detected.
Found 14 outliers among 100 measurements (14.00%)
  9 (9.00%) high mild
  5 (5.00%) high severe
dbscan/1000             time:   [36.916 us 36.989 us 37.069 us]                         
                        change: [-6.6344% -5.4607% -4.3047%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  7 (7.00%) high mild
  3 (3.00%) high severe
dbscan/10000            time:   [375.52 us 376.70 us 378.03 us]                         
                        change: [-11.490% -9.4285% -7.4747%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  4 (4.00%) high mild
  6 (6.00%) high severe
```

Reply:
That looks nice :grin: 
The other optimisation I was considering is adding `cluster_memberships` as a parameter to `find_neighbors` and iterating over both the observations and the memberships using `Zip`, thus returning directly `search_queue`. In this way we avoid having to allocate two vectors (first for neighbors than for the indexes) and we also save a bunch of indexed accesses.

Reply:
Just directly returning the search queue would change the behaviour when it's looking for the first point in the `cluster_id` because the number of neighbours could then drop below `min_points`. There will be a way around that though I'll have a play/think

Reply:
Figured it out, easily enough. Another improvement though not as dramatic this time.

![Screenshot from 2019-12-24 18-50-15](https://user-images.githubusercontent.com/3472518/71423037-4dba2100-267e-11ea-9d88-bd010b5a39de.png)

Reply:
Left a comment which explains better what I was trying to achieve :pray: